### ISSUE TITLE
openpgp-tool --card-info aborts when rendering serial number

### ISSUE THREAD

**Issue by vpelletier** [OpenSC/OpenSC#2404]
First, the card is unusual and experimental: it is [my own openpgp implementation](/vpelletier/python-smartcard-app-openpgp). It certainly has bugs and opensc already helped me correct my understanding of the structure of some responses.

Nevertheless, I think this specific issue does not belong to my side, and may instead be an opensc bug.

Environment and symptoms:

```
$ uname --kernel-version
#1 SMP Debian 5.10.46-1 (2021-06-24)
$ openpgp-tool --version
openpgp-tool - OpenPGP card utility version 0.22.0

Copyright (c) 2012-2020 Peter Marschall <peter@adpm.de>
Licensed under LGPL v2
$ openpgp-tool --card-info
Using reader with a card: Linux Foundation Multifunction Composite Gadget - vincent [python-usb-f-ccid] 00 00
AID:             d2:76:00:01:24:01:03:41:ff:65:a4:9b:68:64:00:00
Version:         3.41
Manufacturer:    unmanaged S/N range
*** buffer overflow detected ***: terminated
Abandon (core dumped)
$ coredumpctl gdb
[...]                
                Stack trace of thread 206287:
                #0  0x00007f18ebc9bce1 __GI_raise (libc.so.6 + 0x3bce1)
                #1  0x00007f18ebc85537 __GI_abort (libc.so.6 + 0x25537)
                #2  0x00007f18ebcde768 __libc_message (libc.so.6 + 0x7e768)
                #3  0x00007f18ebd6d652 __GI___fortify_fail (libc.so.6 + 0x10d652)
                #4  0x00007f18ebd6c050 __GI___chk_fail (libc.so.6 + 0x10c050)
                #5  0x00007f18ebcd7029 _IO_str_chk_overflow (libc.so.6 + 0x77029)
                #6  0x00007f18ebce2ba1 __GI__IO_default_xsputn (libc.so.6 + 0x82ba1)
                #7  0x00007f18ebccaae9 __vfprintf_internal (libc.so.6 + 0x6aae9)
                #8  0x00007f18ebcd70d0 __vsprintf_internal (libc.so.6 + 0x770d0)
                #9  0x00007f18ebd6bb6b ___sprintf_chk (libc.so.6 + 0x10bb6b)
                #10 0x000055e398299307 sprintf (openpgp-tool + 0x4307)
                #11 0x000055e39829968a display_data (openpgp-tool + 0x468a)
                #12 0x000055e398298cac main (openpgp-tool + 0x3cac)
                #13 0x00007f18ebc86d0a __libc_start_main (libc.so.6 + 0x26d0a)
                #14 0x000055e398298d8a _start (openpgp-tool + 0x3d8a)
[...]
(gdb) up
#12 prettify_serialnumber (length=<optimized out>, data=<optimized out>) at openpgp-tool.c:330
330                     sprintf(result, "%08lX", serial);
```

gnupg2 has no problem displaying the serial:
```
$ gpg2 --card-status
[...]
Manufacturer .....: unmanaged S/N range
Serial number ....: A49B6864
[...]
```

APDU extract (`$ openpgp-tool --card-info -vvv`):

```
P:206287; T:0x139745002555456 13:26:46.119 [openpgp-tool] reader-pcsc.c:324:pcsc_transmit: 
Outgoing APDU (7 bytes):
00 CA 00 4F 00 20 00 ...O. .

P:206287; T:0x139745002555456 13:26:46.119 [openpgp-tool] reader-pcsc.c:242:pcsc_internal_transmit: called
P:206287; T:0x139745002555456 13:26:46.137 [openpgp-tool] reader-pcsc.c:333:pcsc_transmit:
Incoming APDU (18 bytes):
D2 76 00 01 24 01 03 41 FF 65 A4 9B 68 64 00 00 .v..$..A.e..hd..
90 00                                           ..
```

I'm not sure why `%08lX` would cause such abort, as the serial is exactly 8 hex digits long.

If you want to reproduce this bug the way I triggered it:

One annoyance in reproducing this bug is that I had to edit `/etc/libccid_Info.plist` for `pcscd` to let me access the reader and card at all (I'm using the very unspecific linux composite gadget usb id: `1D6B:0104`).

Beyond this, getting it running on a GNU/Linux environment with python 3 should be straightforward. Unfortunately it requires privilege elevation in order to setup the usb gadget, which brings its own trust considerations...

```
$ virtualenv -p /usr/bin/python3 "$VENV_DIR"
$ "$VENV_DIR/bin/pip" install smartcard-app-openpgp[ccid]
$ sudo modprobe dummy_hcd
$ sudo modprobe libcomposite
$ sudo "$VENV_DIR/bin/smartcard-openpgp-simple" --username "$USER" --filestorage opencs_test.fs --verbose debug
```

Once the kernel side of the gadget is ready, the gadget itself runs in a process as `$USER`. Any user will do, as long as it can read the `$VENV_DIR` and can write to the path given to `--filestorage`.

This will generate a new card with a new random serial.

**Comment by Jakuje** [comment]
on the first sight, I do not see an issue there. Using gdb you might try to see local variables in the `prettify_serialnumber()` function, unless they are optimized out. If so, you might try to rebuild opensc without optimizations to investigate what is going on there.

**Comment by vpelletier** [comment]
Unfortunately, as you guessed, the values were optimised out. So I built from debian source (but without building a package) with `$ ./configure --disable-optimization` and while it does not crash there is definitely something wrong:

```
$ ./src/tools/openpgp-tool --card-info
Using reader with a card: Linux Foundation Multifunction Composite Gadget - vincent [python-usb-f-ccid] 00 00
AID:             d2:76:00:01:24:01:03:41:ff:65:a4:9b:68:64:00:00
Version:         3.41
Manufacturer:    unmanaged S/N range
Serial number:   FFFFFFFFA49B6864
```

The serial is left-padded (sign-extended ?) to 16 hex digits. The output buffer being `char[15]`, this must be causing a 2 bytes overflow (16th digit and trailing null) in addition of showing those leading digits.

I notice in my original traceback that there is `fortify` mentioned. I'm not sure what this is, and if it is somehow optional this would explain the absence of an abort this time around.

**Comment by vpelletier** [comment]
> (sign-extended ?)

Note sure why, as all involved explicit types are unsigned, but this seems somehow to be the case:

```
$ gdb ./src/tools/.libs/openpgp-tool
[...]
Thread 1 "openpgp-tool" hit Breakpoint 1, prettify_serialnumber (data=0x7fffffffd370 "\244\233hd$\001\003A\377e\244\233hd", length=4) at openpgp-tool.c:326
326             if (data != NULL && length >= 4) {
(gdb) n
328                     unsigned long serial = (unsigned long) (data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3]);
(gdb) n
330                     sprintf(result, "%08lX", serial);
(gdb) print serial
$3 = 18446744072176232548
(gdb) print data[0]
$4 = 164 '\244'
(gdb) print data[0] << 24
$5 = -1543503872
```

**Comment by Jakuje** [comment]
>   (sign-extended ?)

Yeah, that was my initial guess, but I did not see how it happened with all the unsigned types. But I assume some of the left shifts causes the problem.

Using the 32bits ungisnged does not do the trick. It just makes the values smaller to fit the large-enough buffer, but still, the value is not correct. The SN is just 4 bytes that we want to get printed. I think this should be addressed with explicit masking of the bytes as it is done in other parts of the code. I have the minimal reproducer here:
```
$ cat ext.c
/* gcc ext.c -o ext */
#include <stdio.h>
#include <stdint.h>

void main(void)
{
	unsigned char data[] = {0xA4, 0x9B, 0x68, 0x64};
	static char result[15];	/* large enough for even 2*3 digits + separator */
	unsigned long serial = data[0] << 24 & 0xff000000l |
	                       data[1] << 16 & 0x00ff0000l |
	                       data[2] << 8  & 0x0000ff00l |
	                       data[3] << 0  & 0x000000ffl;

	sprintf(result, "%08lX", serial);
	printf("%s\n", result);
}
$ gcc ext.c -o ext
$ ./ext 
A49B6864
```
We have quite extensive fuzzing tests of the drivers, but these things slip as they are present only in the tools. They do not pose so large threat, but I understand that here is a room for improvement.

**Comment by dengert** [comment]
according to this: 
https://stackoverflow.com/questions/9860538/unexpected-c-c-bitwise-shift-operators-outcome  Shifting a unsigned char more then its size(8 bits))  is undefined in C and C++.
```
$ cat ext.c
/* gcc ext.c -o ext */
#include <stdio.h>
#include <stdint.h>

void main(void)
{
	unsigned char data[] = {0xA4, 0x9B, 0x68, 0x64};
	static char result[15];	/* large enough for even 2*3 digits + separator */
	unsigned long serial = data[0] << 24 & 0xff000000l |
	                       data[1] << 16 & 0x00ff0000l |
	                       data[2] << 8  & 0x0000ff00l |
	                       data[3] << 0  & 0x000000ffl;
        unsigned long serial2 = (unsigned long) (data[0] << 24 | data[1] << 16 | data[2] << 8 | data[3]);
        unsigned long serial3 = (unsigned int)(data[0] << 24) | (unsigned int)(data[1] << 16) | (unsigned int)(data[2] << 8) | (unsigned int)(data[3]);
	
	printf("sizeof(unsigned long):%ld sizeof(unsigned int):%ld sizeof(unsigned char):%ld\n", sizeof(unsigned long), sizeof(unsigned int), sizeof(unsigned char));

	sprintf(result, "%08lX", serial);
	printf("serial %s\n", result);

	sprintf(result, "%08lX", serial2);
	printf("serial2:%s\n", result);

	sprintf(result, "%08lX", serial3);
	printf("serial3:%s\n", result);
}
$ gcc ext.c -o ext
$ ./ext
sizeof(unsigned long):8 sizeof(unsigned int):4 sizeof(unsigned char):1
serial A49B6864
serial2:FFFFFFFFA49B6864
serial3:A49B6864
$gcc --version
gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
```

So "serial"  with "&" bit mask makes best sense as should work on any machine.

**Comment by vpelletier** [comment]
> They do not pose so large threat

This was my understanding as well when I opened this issue. I rather saw this as getting in the way of me maybe getting to the next APDU and finding out that I got some other response encoding wrong :) . But with the debug build I actually see there is no next APDU. So I'll test other functions to find if I can find the next bug (...hopefully on my side).

> So "serial" with "&" bit mask makes best sense as should work on any machine.

As a clueless bystander: what about `sprintf("%02X%02X%02X%02X", data[0], data[1], data[2], data[3])` ?

**Comment by dengert** [comment]
Yes, `sprintf("%02X%02X%02X%02X", data[0], data[1], data[2], data[3]);` works.

added these to the ext.c: 
```
         sprintf(result, "%02X%02X%02X%02X", data[0], data[1], data[2], data[3]);
         printf("using %%02X%%02X%%02X%%02X result:%s\n", result); 
```
with output:
```
using %02X%02X%02X%02X result:A49B6864
```

**Comment by Jakuje** [comment]
>  https://stackoverflow.com/questions/9860538/unexpected-c-c-bitwise-shift-operators-outcome Shifting a unsigned char more then its size(8 bits)) is undefined in C and C++.

I am not sure if the `u8`/`unsigned char` is not automatically promoted to some larger types here, clearly from the code above, it looks like it works for some reason, when the sign-extension it is correctly masked.

> As a clueless bystander: what about `sprintf("%02X%02X%02X%02X", data[0], data[1], data[2], data[3])` ?

I like this option the best. It is simple and it works. Do you want to submit a PR with the fix?

**Comment by vpelletier** [comment]
> Do you want to submit a PR with the fix?

Done: #2405 

I have no idea how much I should detail the commit message, and I am definitely not able to explain how the cast/shift is going wrong with any semblance of certainty. Please feel free to edit as appropriate.

The output is now:
```
$ ./src/tools/.libs/openpgp-tool --card-info
Using reader with a card: Linux Foundation Multifunction Composite Gadget - vincent [python-usb-f-ccid] 00 00
AID:             d2:76:00:01:24:01:03:41:ff:65:a4:9b:68:64:00:00
Version:         3.41
Manufacturer:    unmanaged S/N range
Serial number:   A49B6864
```

**Comment by Jakuje** [comment]
The #2405 is merged so closing also this issue. Thank you for the report and the work on the fix (and tests!).
